src/aspire/utils/relion_interop.py

```diff
@@ -50,42 +50,130 @@
+        """
+
+        self.data_block_name = ""
+        self.relion_version = ""
+
+        # validate 3.0 STAR file
+        if len(self.blocks) == 1:
+            data_block = self.get_block_by_index(0)
+            columns = data_block.columns.to_list()
+            if not any(
+                col in columns
+                for col in ["_rlnImageName", "_rlnMicrographName", "_rlnMovieName"]
+            ):
+                raise ValueError(
+                    f"{self.filepath} does not contain Relion data columns."
+                )
+
+            self.relion_version = "3.0"
+            self.data_block = data_block
+
+        # validate 3.1 STAR file
+        if len(self.blocks) == 2:
+            # must have an optics block
+            if "optics" not in self.blocks.keys():
+                raise ValueError(f"{self.filepath} does not contain an optics block.")
+
+            # find type of data
+            for name in self.blocks.keys():
+                if name in ["particles", "micrographs", "movies"]:
+                    self.data_block_name = name
+                    break
+            if not self.data_block_name:
+                raise ValueError(
+                    f"{self.filepath} does not contain a block identifying particle, ",
+                    "micrograph, or movie data.",
+                )
+
+            data_block = self[self.data_block_name]
+            # lastly, data block must contain a column identifying the type of data as well
+            columns = data_block.columns.to_list()
+            if not any(
+                col in columns
+                for col in ["_rlnImageName", "_rlnMicrographName", "_rlnMovieName"]
+            ):
+                raise ValueError(
+                    f"{self.filepath} data block does not contain Relion data columns."
+                )
+
+            self.relion_version = "3.1"
+            self.data_block = data_block
+            self.optics_block = self["optics"]
+
+        if not self.relion_version:
+            raise ValueError(
+                f"Cannot interpret {self.filepath} as a valid RELION STAR file."
+            )
+
+    def convert_dtypes(self):
+        """
+        For data blocks that are Pandas DataFrames, convert known Relion columns
+        to sensible types.
+        """
         _blocks = OrderedDict()
         for block_name, block in self.blocks.items():
-            _blocks[block_name] = df_to_relion_types(block)
+            # generic StarFile blocks can be DataFrame or dict
+            if not isinstance(block, dict):
+                _blocks[block_name] = df_to_relion_types(block)
         self.blocks = _blocks
 
-
-class Relion31StarFile(Relion30StarFile):
-    def __init__(self, filepath):
-        super().__init__(filepath)
-        self.optics_block = self.get_block_by_index(0)
-        self.data_block = self.get_block_by_index(1)
-
-    def apply_optics_block(self):
+    def get_data_block(self):
```

Comment:
Would probably rename this `get_merged_data_block`. Otherwise it's not clear how this differs from the `data_block` attribute.

Reply:
Renamed